stateresolutionv2.go

```diff
@@ -0,0 +1,459 @@
+func (r *stateResolverV2) Create() (*Event, error) {
+	return r.resolvedCreate, nil
+}
+
+// PowerLevels implements AuthEventProvider
+func (r *stateResolverV2) PowerLevels() (*Event, error) {
+	return r.resolvedPowerLevels, nil
+}
+
+// JoinRules implements AuthEventProvider
+func (r *stateResolverV2) JoinRules() (*Event, error) {
+	return r.resolvedJoinRules, nil
+}
+
+// ThirdPartyInvite implements AuthEventProvider
+func (r *stateResolverV2) ThirdPartyInvite(key string) (*Event, error) {
+	return r.resolvedThirdPartyInvites[key], nil
+}
+
+// Member implements AuthEventProvider
+func (r *stateResolverV2) Member(key string) (*Event, error) {
+	return r.resolvedMembers[key], nil
+}
+
+// ResolveStateConflicts takes a list of state events with conflicting state
+// keys and works out which event should be used for each state event. This
+// function returns the resolved state, including unconflicted state events.
+func ResolveStateConflictsV2(conflicted, unconflicted []Event, authEvents []Event) []Event {
+	var conflictedPowerLevels []Event
+	var conflictedOthers []Event
+	r := stateResolverV2{
+		authEventMap:              eventMapFromEvents(authEvents),
+		resolvedThirdPartyInvites: make(map[string]*Event),
+		resolvedMembers:           make(map[string]*Event),
+	}
+
+	// Separate out power level events from the rest of the events. This is
+	// necessary because we perform topological ordering of the power events
+	// separately, and then the mainline ordering of all other events depends
+	// on that power level ordering.
+	for _, p := range conflicted {
+		if p.Type() == MRoomPowerLevels {
+			conflictedPowerLevels = append(conflictedPowerLevels, p)
+		} else {
+			conflictedOthers = append(conflictedOthers, p)
+		}
+	}
+
+	// Start with the unconflicted events by ordering them topologically and then
+	// authing them. The successfully authed events will form the initial partial
+	// state.
+	unconflicted = r.reverseTopologicalOrdering(unconflicted)
+	r.authAndApplyEvents(unconflicted)
+
+	// Then order the conflicted power level events topologically and then also
+	// auth those too. The successfully authed events will be layered on top of
+	// the partial state.
+	conflictedPowerLevels = r.reverseTopologicalOrdering(conflictedPowerLevels)
+	r.authAndApplyEvents(conflictedPowerLevels)
+
+	// Then generate the mainline of power level events, order the remaining state
+	// events based on the mainline ordering and auth those too. The successfully
+	// authed events are also layered on top of the partial state.
+	r.powerLevelMainline = r.createPowerLevelMainline()
+	r.authAndApplyEvents(r.mainlineOrdering(conflictedOthers))
+
+	// Finally we will reapply the original set of unconflicted events onto the
+	// partial state, just in case any of these were overwritten by pulling in
+	// auth events in the previous two steps, and that gives us our final resolved
+	// state.
+	r.authAndApplyEvents(unconflicted)
```

Comment:
You're right about this one though - I think we should just be able to reapply the original set of authed unconflicted events rather than authing them again.